addons/account_ubl/models/account_invoice.py

```diff
@@ -0,0 +1,193 @@
+# -*- coding: utf-8 -*-
+
+from odoo import models, fields, api, tools, _
+from odoo.exceptions import UserError
+
+import copy
+import base64
+
+from lxml import etree
+from StringIO import StringIO
+from tempfile import NamedTemporaryFile
+from PyPDF2 import PdfFileWriter, PdfFileReader
+
+UBL_NAMESPACES = {
+    'cac': '{urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2}',
+    'cbc': '{urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2}',
+}
+
+''' UBL_NS_REFACTORING is used to restore the right namespaces inside the xml.
+This method is necessary when using Qweb because the <t></t> elements cause some 
+troubles when some namespaces are specified to etree. So, to avoid this problem,
+the namespaces are encoded with a prefix in each tag and can be replaced as a namespace definition
+after the Qweb rendering. In this case, the prefix is 'cbc__' or 
+'cac__' and is replaced by 'cbc:' or 'cac:' respectively.
+'''
+UBL_NS_REFACTORING = {
+    'cbc__': 'cbc',
+    'cac__': 'cac',
+}
+
+class AccountInvoice(models.Model):
+    _name = 'account.invoice'
+    _inherit = 'account.invoice'
+
+    ubl_doc_ref = fields.Binary(
+        string='The document reference content as binary',
+        help='contains the report as .pdf to embed in the document reference node for e-fff (UBL)',
+        readonly=True)
+
+    @api.multi
+    def prepare_pdf(self, values=None):
+        # We need to save this content as field because the report is generated only once.
+        if values:
+            for record in self:
+                if not record.ubl_doc_ref:
+                    record.ubl_doc_ref = values['datas']
+        return super(AccountInvoice, self).prepare_pdf(values=values)
+
+    @api.multi
+    def get_edi_filenames_BE(self):
+        ''' The filename for e-fff doesn't seem to have an official pattern, so
+        this filename is defined arbitrarily.
+        '''
+        return [('%s-UBL-Invoice-2.1.xml' % self.number).replace('/', '')]
+
+    @api.model
+    def generate_edi_BE(self):
+        ''' Generate the EDI file for belgium,
+        following the UBL protocol e-fff
+        '''
+        filename = self.get_edi_filenames_BE()[0]
+        values = self._create_ubl_values_BE()
+        qweb = self.env['ir.qweb']
+        content = qweb.render('account_ubl.ubl_invoice_e_fff', values=values)
+        
+        # TEMP: refactoring namespaces
+        for key, value in UBL_NS_REFACTORING.items():
+            content = content.replace(key, value + ':')
```

Comment:
What about use `content = content.replace('<' + key, '<' + value + ':')` in order to avoid replace a content value instead of a tag name?